src/osd/PrimaryLogPG.cc

```diff
@@ -2570,10 +2560,23 @@ int PrimaryLogPG::do_manifest_flush(OpRequestRef op, ObjectContextRef obc, Flush
     unsigned flags = CEPH_OSD_FLAG_IGNORE_CACHE | CEPH_OSD_FLAG_IGNORE_OVERLAY |
 		     CEPH_OSD_FLAG_RWORDERED;
     tgt_length = chunk_data.length();
-    if (pg_pool_t::fingerprint_t fp_algo = pool.info.get_fingerprint_type();
-	iter->second.has_reference() &&
-	fp_algo != pg_pool_t::TYPE_FINGERPRINT_NONE) {
-      object_t fp_oid = [fp_algo, &chunk_data]() -> string {
+
+    /* 
+     * TODO:
+     * set_chunk will not imply that flush eventually re-write 
+     * the chunk if it becomes overwritten. So, we need to remove this part
+     * entirely and rework the dedup procedure based on thw following scenarios.
```

Comment:
I think you should actually remove the code.  I don't think that any of the code in do_manifest_flush makes sense.  It iterates over exactly the chunks that don't need to be "flushed" and doesn't perform the CDC yet.  I don't see a reason to leave any of it here -- you should probably just return an error in the event that start_flush is invoked on a manifest object.

Also, this TODO refers to a misunderstanding we hope readers won't have in the first place, please replace it with:

```
TODO: "flush" for a manifest object means re-running the CDC algorithm on the portions of the object that are not currently dedup'd (not in the manifest chunk_map) and re-deduping the resulting chunks.  Adding support for that operation here is future work.
```

Reply:
Removing this code here means some test cases including unit and stress test also need to be deleted because they make use of current flush. Do you agree on this?

Reply:
When called on an object with existing dedup chunks, this implementation appears to leak one refcount for every existing chunk (up to get_copy_chunk_size()) and appears to have no other effect.  Is that correct?

I think that if the above is correct I don't really consider the tests that cover it to be valuable.

Reply:
> When called on an object with existing dedup chunks, this implementation appears to leak one refcount for every existing chunk (up to get_copy_chunk_size()) and appears to have no other effect. Is that correct?

There is one thing to note. It flushs (read the object and send the content) the content to the target object in the low tier.

Reply:
Perhaps I have missed something.

From what I can tell, it iterates over the object's current object_info.chunk_map.  For each (offset, len) region already in chunk_map, it reads (offset, len) from the object, computes a fingerprint (foo), and then invokes chunk_create_or_get_ref on the object name resulting from the fingerprint (foo).

I think the above is what you mean when you say that it flushes the data at (offset, len) to the target object (foo) in the lower tier.

However, as mentioned above, (offset, len) is already in the chunk_map.  Because it is already in the map, it already has a reference to a lower tier object <x> based on the fingerprint computed at the time from the data at (offset, len) .  Because we now remove chunk_map entries upon modification, the fact that that entry still exists means that it cannot have changed.  Therefore, the fingerprint <x> that already existed *must equal foo already*.  Thus, chunk_create_or_get_ref has exactly one effect: it unnecessarily increases the refcount on the dedup target the object already had.

In other words, do_manifest_flush here iterates over exactly the object extents that do not need to be flushed and increments their refcounts incorrectly.

Reply:
Right, it unnecessarily increases the refcount. If a new flush is implemented, this code is not necessary. I mean
you also need to consider not only counting but also flush.
I already added a commit that address your comment.

Reply:
@myoungwon Please read my whole comment.  I don't mean that it performs a flush and also leaks a reference.  I mean that the only observable effect of this method is that the chunks up to get_copy_chunk_size() have their refcounts increased by one.  The target objects selected for each chunk are identical to the ones that must already exist, so the operation is otherwise a noop.

Reply:
In other words, unless I'm misunderstanding something, if this method returned 0 and did literally nothing else, it would have the same effect, but wouldn't leak references.

Reply:
Never mind my comment above (I mean you also need to consider not only counting but also flush). 
I understand what you say.